crates/biome_js_analyze/src/lint/nursery/no_useless_undefined.rs

```diff
@@ -0,0 +1,303 @@
+use biome_analyze::{
+    context::RuleContext, declare_lint_rule, ActionCategory, Ast, FixKind, Rule, RuleDiagnostic,
+    RuleSource,
+};
+use biome_console::markup;
+use biome_js_factory::make::{self, js_function_body};
+use biome_js_syntax::{
+    AnyJsExpression, AnyJsFunctionBody, JsArrayBindingPatternElement, JsArrowFunctionExpression,
+    JsFormalParameter, JsObjectBindingPatternShorthandProperty, JsReturnStatement,
+    JsVariableStatement, JsYieldArgument, T,
+};
+use biome_rowan::{declare_node_union, AstNode, BatchMutationExt, TextRange};
+
+use crate::JsRuleAction;
+
+declare_lint_rule! {
+    /// Disallow the use of useless `undefined`.
+    ///
+    /// `undefined` is the default value for new variables, parameters, return statements, etc., so specifying it doesn't make any difference.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// let foo = undefined;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const {foo = undefined} = bar;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const noop = () => undefined;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// function foo() {
+    ///    return undefined;
+    /// }
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// function* foo() {
+    ///   yield undefined;
+    /// }
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// function foo(bar = undefined) {}
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// function foo({bar = undefined}) {}
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// foo(undefined);
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// let foo;
+    /// const {foo} = bar;
+    /// function foo() {
+    ///   return;
+    /// }
+    /// function* foo() {
+    ///   yield;
+    /// }
+    /// function foo(bar) {}
+    /// function foo({bar}) {}
+    /// foo();
+    /// ```
+    ///
+    pub NoUselessUndefined {
+        version: "next",
+        name: "noUselessUndefined",
+        language: "js",
+        fix_kind: FixKind::Safe,
+        sources: &[RuleSource::EslintUnicorn("no-useless-undefined")],
+        recommended: false,
+    }
+}
+
+declare_node_union! {
+    pub AnyUndefinedNode = JsVariableStatement
+        | JsObjectBindingPatternShorthandProperty
+        | JsYieldArgument
+        | JsReturnStatement
+        | JsArrayBindingPatternElement
+        | JsArrowFunctionExpression
+        | JsFormalParameter
+}
+
+fn find_undefined_range(expr: Option<&AnyJsExpression>) -> Option<TextRange> {
+    let ident = expr?.as_js_reference_identifier()?;
+    if ident.is_undefined() {
+        Some(ident.range())
+    } else {
+        None
+    }
+}
+
+pub struct RuleState {
+    binding_text: Option<Box<str>>,
```

Comment:
To avoid a string allocation we could use `TokenText`.
```suggestion
    binding_text: Option<TokenText>,
```